consensus/src/marshal/mod.rs

```diff
@@ -407,13 +428,15 @@ mod tests {
 
                 // Finalize block by all validators
                 let fin = make_finalization(proposal, &schemes, QUORUM);
-                for actor in actors.iter_mut() {
+                for (i, actor) in actors.iter_mut().enumerate() {
                     // Always finalize 1) the last block in each epoch 2) the last block in the chain.
-                    // Otherwise, finalize randomly.
-                    if height == NUM_BLOCKS
+                    //
+                    // Otherwise:
+                    // - If `always_finalize` is set, finalize every block for all actors other than the proposer.
+                    // - If `always_finalize` is not set, finalize randomly with 20% chance.
+                    if ((always_finalize && i != actor_index) || context.gen_bool(0.2))
```

Comment:
Instead of `always_finalize`, we could still periodically finalize but just ensure most participants see it